File: src/amazonq/parser.py
"""Amazon Q Event Stream 解析器"""
import json
import struct
import logging
from typing import Optional, Dict, Any, AsyncIterator

logger = logging.getLogger(__name__)

class EventStreamParser:
    """AWS Event Stream 二进制格式解析器"""

    @staticmethod
    def parse_headers(headers_data: bytes) -> Dict[str, str]:
        """
        解析事件流头部

        Args:
            headers_data: 头部二进制数据

        Returns:
            解析后的头部字典
        """
        headers = {}
        offset = 0

        while offset < len(headers_data):
            if offset >= len(headers_data):
                break
            name_length = headers_data[offset]
            offset += 1

            if offset + name_length > len(headers_data):
                break
            name = headers_data[offset:offset + name_length].decode('utf-8')
            offset += name_length

            if offset >= len(headers_data):
                break
            value_type = headers_data[offset]
            offset += 1

            if offset + 2 > len(headers_data):
                break
            value_length = struct.unpack('>H', headers_data[offset:offset + 2])[0]
            offset += 2

            if offset + value_length > len(headers_data):
                break

            if value_type == 7:
                value = headers_data[offset:offset + value_length].decode('utf-8')
            else:
                value = headers_data[offset:offset + value_length]

            offset += value_length
            headers[name] = value

        return headers

    @staticmethod
    def parse_message(data: bytes) -> Optional[Dict[str, Any]]:
        """
        解析单个 Event Stream 消息

        Args:
            data: 消息二进制数据

        Returns:
            解析后的消息字典
        """
        try:
            if len(data) < 16:
                return None

            total_length = struct.unpack('>I', data[0:4])[0]
            headers_length = struct.unpack('>I', data[4:8])[0]

            if len(data) < total_length:
                logger.warning(f"Incomplete message: expected {total_length} bytes, got {len(data)}")
                return None

            headers_data = data[12:12 + headers_length]
            headers = EventStreamParser.parse_headers(headers_data)

            payload_start = 12 + headers_length
            payload_end = total_length - 4
            payload_data = data[payload_start:payload_end]

            payload = None
            if payload_data:
                try:
                    payload = json.loads(payload_data.decode('utf-8'))
                except (json.JSONDecodeError, UnicodeDecodeError):
                    payload = payload_data

            return {
                'headers': headers,
                'payload': payload,
                'total_length': total_length
            }

        except Exception as e:
            logger.error(f"Failed to parse message: {e}", exc_info=True)
            return None

    @staticmethod
    async def parse_stream(byte_stream: AsyncIterator[bytes]) -> AsyncIterator[Dict[str, Any]]:
        """
        解析字节流并提取事件

        Args:
            byte_stream: 异步字节流

        Yields:
            解析后的事件消息
        """
        buffer = bytearray()

        async for chunk in byte_stream:
            buffer.extend(chunk)

            while len(buffer) >= 12:
                try:
                    total_length = struct.unpack('>I', buffer[0:4])[0]
                except struct.error:
                    break

                if len(buffer) < total_length:
                    break

                message_data = bytes(buffer[:total_length])
                buffer = buffer[total_length:]

                message = EventStreamParser.parse_message(message_data)
                if message:
                    yield message

def extract_event_info(message: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """
    从解析后的消息中提取事件信息

    Args:
        message: 解析后的消息

    Returns:
        事件信息字典
    """
    headers = message.get('headers', {})
    payload = message.get('payload')

    event_type = headers.get(':event-type') or headers.get('event-type')
    content_type = headers.get(':content-type') or headers.get('content-type')
    message_type = headers.get(':message-type') or headers.get('message-type')

    return {
        'event_type': event_type,
        'content_type': content_type,
        'message_type': message_type,
        'payload': payload
    }

def _sse_format(event_type: str, data: Dict[str, Any]) -> str:
    """
    格式化 SSE 事件

    Args:
        event_type: 事件类型
        data: 事件数据

    Returns:
        SSE 格式字符串
    """
    json_data = json.dumps(data, ensure_ascii=False)
    return f"event: {event_type}\ndata: {json_data}\n\n"

def build_message_start(conversation_id: str, model: str = "claude-sonnet-4.5", input_tokens: int = 0) -> str:
    """构建 message_start SSE 事件"""
    data = {
        "type": "message_start",
        "message": {
            "id": conversation_id,
            "type": "message",
            "role": "assistant",
            "content": [],
            "model": model,
            "stop_reason": None,
            "stop_sequence": None,
            "usage": {"input_tokens": input_tokens, "output_tokens": 0}
        }
    }
    return _sse_format("message_start", data)

def build_content_block_start(index: int, block_type: str = "text") -> str:
    """构建 content_block_start SSE 事件"""
    data = {
        "type": "content_block_start",
        "index": index,
        "content_block": {"type": block_type, "text": ""} if block_type == "text" else {"type": block_type}
    }
    return _sse_format("content_block_start", data)

def build_content_block_delta(index: int, text: str) -> str:
    """构建 content_block_delta SSE 事件（文本）"""
    data = {
        "type": "content_block_delta",
        "index": index,
        "delta": {"type": "text_delta", "text": text}
    }
    return _sse_format("content_block_delta", data)

def build_content_block_stop(index: int) -> str:
    """构建 content_block_stop SSE 事件"""
    data = {
        "type": "content_block_stop",
        "index": index
    }
    return _sse_format("content_block_stop", data)

def build_ping() -> str:
    """构建 ping SSE 事件"""
    data = {"type": "ping"}
    return _sse_format("ping", data)

def build_message_stop(input_tokens: int, output_tokens: int, stop_reason: Optional[str] = None) -> str:
    """构建 message_delta 和 message_stop SSE 事件"""
    delta_data = {
        "type": "message_delta",
        "delta": {"stop_reason": stop_reason or "end_turn", "stop_sequence": None},
        "usage": {"output_tokens": output_tokens}
    }
    delta_event = _sse_format("message_delta", delta_data)

    stop_data = {
        "type": "message_stop"
    }
    stop_event = _sse_format("message_stop", stop_data)

    return delta_event + stop_event

def build_tool_use_start(index: int, tool_use_id: str, tool_name: str) -> str:
    """构建 tool_use content_block_start SSE 事件"""
    data = {
        "type": "content_block_start",
        "index": index,
        "content_block": {
            "type": "tool_use",
            "id": tool_use_id,
            "name": tool_name,
            "input": {}
        }
    }
    return _sse_format("content_block_start", data)

def build_tool_use_input_delta(index: int, input_json_delta: str) -> str:
    """构建 tool_use input_json_delta SSE 事件"""
    data = {
        "type": "content_block_delta",
        "index": index,
        "delta": {
            "type": "input_json_delta",
            "partial_json": input_json_delta
        }
    }
    return _sse_format("content_block_delta", data)

File: internal/amazonq/parser.go
package amazonq

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
)

// EventStreamMessage 表示事件流中的单个消息
type EventStreamMessage struct {
	Headers     map[string]string
	Payload     interface{}
	TotalLength uint32
}

// EventInfo 存储解析后的事件信息
type EventInfo struct {
	EventType   string
	ContentType string
	MessageType string
	Payload     interface{}
}

// SSEEvent 表示 Server-Sent Events 事件结构
type SSEEvent struct {
	Event string
	Data  interface{}
}

// ParseHeaders 解析事件流消息的头部数据
// 参数 headersData 为头部二进制数据
// 返回解析后的头部键值对映射
func ParseHeaders(headersData []byte) map[string]string {
	headers := make(map[string]string)
	offset := 0

	for offset < len(headersData) {
		if offset >= len(headersData) {
			break
		}
		nameLength := int(headersData[offset])
		offset++

		if offset+nameLength > len(headersData) {
			break
		}
		name := string(headersData[offset : offset+nameLength])
		offset += nameLength

		if offset >= len(headersData) {
			break
		}
		valueType := headersData[offset]
		offset++

		if offset+2 > len(headersData) {
			break
		}
		valueLength := binary.BigEndian.Uint16(headersData[offset : offset+2])
		offset += 2

		if offset+int(valueLength) > len(headersData) {
			break
		}

		var value string
		if valueType == 7 {
			value = string(headersData[offset : offset+int(valueLength)])
		} else {
			value = string(headersData[offset : offset+int(valueLength)])
		}

		offset += int(valueLength)
		headers[name] = value
	}

	return headers
}

// ParseMessage 解析单个 Event Stream 消息
// 参数 data 为完整的消息二进制数据
// 返回解析后的消息结构和可能的错误
func ParseMessage(data []byte) (*EventStreamMessage, error) {
	if len(data) < 16 {
		return nil, fmt.Errorf("incomplete message: data too short")
	}

	totalLength := binary.BigEndian.Uint32(data[0:4])
	headersLength := binary.BigEndian.Uint32(data[4:8])

	if len(data) < int(totalLength) {
		return nil, fmt.Errorf("incomplete message: expected %d bytes, got %d", totalLength, len(data))
	}

	headersData := data[12 : 12+headersLength]
	headers := ParseHeaders(headersData)

	payloadStart := 12 + headersLength
	payloadEnd := totalLength - 4
	payloadData := data[payloadStart:payloadEnd]

	var payload interface{}
	if len(payloadData) > 0 {
		err := json.Unmarshal(payloadData, &payload)
		if err != nil {
			payload = string(payloadData)
		}
	}

	return &EventStreamMessage{
		Headers:     headers,
		Payload:     payload,
		TotalLength: totalLength,
	}, nil
}

// ParseStream 从字节流中解析事件并发送到通道
// 参数 reader 为字节流读取器
// 参数 eventChan 为事件输出通道
// 返回可能的错误
func ParseStream(reader io.Reader, eventChan chan<- *EventStreamMessage) error {
	defer close(eventChan)

	buffer := make([]byte, 0)
	chunk := make([]byte, 4096)

	for {
		n, err := reader.Read(chunk)
		if n > 0 {
			buffer = append(buffer, chunk[:n]...)

			for len(buffer) >= 12 {
				if len(buffer) < 4 {
					break
				}

				totalLength := binary.BigEndian.Uint32(buffer[0:4])

				if len(buffer) < int(totalLength) {
					break
				}

				messageData := buffer[:totalLength]
				buffer = buffer[totalLength:]

				message, parseErr := ParseMessage(messageData)
				if parseErr != nil {
					continue
				}

				eventChan <- message
			}
		}

		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}
	}

	return nil
}

// ExtractEventInfo 从解析后的消息中提取事件信息
// 参数 message 为解析后的消息结构
// 返回提取的事件信息
func ExtractEventInfo(message *EventStreamMessage) *EventInfo {
	headers := message.Headers

	eventType := headers[":event-type"]
	if eventType == "" {
		eventType = headers["event-type"]
	}

	contentType := headers[":content-type"]
	if contentType == "" {
		contentType = headers["content-type"]
	}

	messageType := headers[":message-type"]
	if messageType == "" {
		messageType = headers["message-type"]
	}

	return &EventInfo{
		EventType:   eventType,
		ContentType: contentType,
		MessageType: messageType,
		Payload:     message.Payload,
	}
}

// FormatSSE 将事件格式化为 Server-Sent Events 格式
// 参数 eventType 为事件类型
// 参数 data 为事件数据
// 返回 SSE 格式字符串
func FormatSSE(eventType string, data interface{}) string {
	jsonData, _ := json.Marshal(data)
	return fmt.Sprintf("event: %s\ndata: %s\n\n", eventType, string(jsonData))
}

// BuildMessageStart 构建 message_start SSE 事件
// 参数 conversationID 为会话 ID
// 参数 model 为模型名称
// 参数 inputTokens 为输入 token 数量
// 返回 SSE 格式的事件字符串
func BuildMessageStart(conversationID string, model string, inputTokens int) string {
	data := map[string]interface{}{
		"type": "message_start",
		"message": map[string]interface{}{
			"id":            conversationID,
			"type":          "message",
			"role":          "assistant",
			"content":       []interface{}{},
			"model":         model,
			"stop_reason":   nil,
			"stop_sequence": nil,
			"usage": map[string]int{
				"input_tokens":  inputTokens,
				"output_tokens": 0,
			},
		},
	}
	return FormatSSE("message_start", data)
}

// BuildContentBlockStart 构建 content_block_start SSE 事件
// 参数 index 为内容块索引
// 参数 blockType 为内容块类型（如 "text", "thinking" 或 "tool_use"）
// 返回 SSE 格式的事件字符串
func BuildContentBlockStart(index int, blockType string) string {
	contentBlock := map[string]interface{}{"type": blockType}
	if blockType == "text" || blockType == "thinking" {
		contentBlock["text"] = ""
	}

	data := map[string]interface{}{
		"type":          "content_block_start",
		"index":         index,
		"content_block": contentBlock,
	}
	return FormatSSE("content_block_start", data)
}

// BuildContentBlockDelta 构建 content_block_delta SSE 事件（文本增量）
// 参数 index 为内容块索引
// 参数 text 为增量文本内容
// 返回 SSE 格式的事件字符串
func BuildContentBlockDelta(index int, text string) string {
	data := map[string]interface{}{
		"type":  "content_block_delta",
		"index": index,
		"delta": map[string]string{
			"type": "text_delta",
			"text": text,
		},
	}
	return FormatSSE("content_block_delta", data)
}

// BuildContentBlockStop 构建 content_block_stop SSE 事件
// 参数 index 为内容块索引
// 返回 SSE 格式的事件字符串
func BuildContentBlockStop(index int) string {
	data := map[string]interface{}{
		"type":  "content_block_stop",
		"index": index,
	}
	return FormatSSE("content_block_stop", data)
}

// BuildPing 构建 ping SSE 事件，用于保持连接活跃
// 返回 SSE 格式的 ping 事件字符串
func BuildPing() string {
	data := map[string]string{"type": "ping"}
	return FormatSSE("ping", data)
}

// BuildMessageStop 构建 message_delta 和 message_stop SSE 事件
// 参数 inputTokens 为输入 token 数量
// 参数 outputTokens 为输出 token 数量
// 参数 stopReason 为停止原因（可为 nil）
// 返回组合的 SSE 格式事件字符串
func BuildMessageStop(inputTokens int, outputTokens int, stopReason *string) string {
	reason := "end_turn"
	if stopReason != nil {
		reason = *stopReason
	}

	deltaData := map[string]interface{}{
		"type": "message_delta",
		"delta": map[string]interface{}{
			"stop_reason":   reason,
			"stop_sequence": nil,
		},
		"usage": map[string]int{
			"output_tokens": outputTokens,
		},
	}
	deltaEvent := FormatSSE("message_delta", deltaData)

	stopData := map[string]string{"type": "message_stop"}
	stopEvent := FormatSSE("message_stop", stopData)

	return deltaEvent + stopEvent
}

// BuildToolUseStart 构建 tool_use 类型的 content_block_start SSE 事件
// 参数 index 为内容块索引
// 参数 toolUseID 为工具使用 ID
// 参数 toolName 为工具名称
// 返回 SSE 格式的事件字符串
func BuildToolUseStart(index int, toolUseID string, toolName string) string {
	data := map[string]interface{}{
		"type":  "content_block_start",
		"index": index,
		"content_block": map[string]interface{}{
			"type":  "tool_use",
			"id":    toolUseID,
			"name":  toolName,
			"input": map[string]interface{}{},
		},
	}
	return FormatSSE("content_block_start", data)
}

// BuildToolUseInputDelta 构建 tool_use 的 input_json_delta SSE 事件
// 参数 index 为内容块索引
// 参数 inputJSONDelta 为 JSON 增量内容
// 返回 SSE 格式的事件字符串
func BuildToolUseInputDelta(index int, inputJSONDelta string) string {
	data := map[string]interface{}{
		"type":  "content_block_delta",
		"index": index,
		"delta": map[string]string{
			"type":         "input_json_delta",
			"partial_json": inputJSONDelta,
		},
	}
	return FormatSSE("content_block_delta", data)
}
